File: compiler/nodes/types/block.go
package types

import (
	"wipple/database"
	"wipple/syntax"
	"wipple/typecheck"
	"wipple/visit"
)

type BlockTypeNode struct {
	Output database.Node
	Facts  *database.Facts
}

func (node *BlockTypeNode) GetFacts() *database.Facts {
	return node.Facts
}

func ParseBlockType(parser *syntax.Parser) (*BlockTypeNode, *syntax.Error) {
	span := parser.Spanned()

	_, err := parser.Token("LeftBrace")
	if err != nil {
		return nil, err
	}

	output, err := ParseTypeElement(parser)
	if err != nil {
		return nil, err
	}

	_, err = parser.Token("RightBrace")
	if err != nil {
		return nil, err
	}

	return &BlockTypeNode{
		Output: output,
		Facts:  database.NewFacts(span()),
	}, nil
}

func (node *BlockTypeNode) Visit(visitor *visit.Visitor) {
	visitType(visitor, node)

	visitor.Visit(node.Output)
	visitor.Constraint(typecheck.TypeConstraint(node, typecheck.BlockType(node.Output)))
}

File: compiler/src/nodes/types/block.rs
use crate::{
    codegen::{Codegen, CodegenCtx, CodegenError},
    database::{Node, NodeRef},
    nodes::{parse_type_element, visit_type},
    syntax::{ParseError, Parser, TokenKind},
    typecheck::TypeConstraint,
    visit::{Visit, Visitor},
};

#[derive(Debug)]
pub struct BlockTypeNode {
    pub output: NodeRef,
}

impl Node for BlockTypeNode {}

pub fn parse_block(parser: &mut Parser<'_>) -> Result<BlockTypeNode, ParseError> {
    parser.token(TokenKind::LeftBrace)?;
    let output = parse_type_element(parser)?;
    parser.token(TokenKind::RightBrace)?;

    Ok(BlockTypeNode { output })
}

impl Visit for BlockTypeNode {
    fn visit(&self, node: &NodeRef, visitor: &mut Visitor<'_>) {
        visit_type(node, visitor);

        visitor.visit(&self.output);
        visitor.edge(&self.output, node, "output");

        visitor.constraint(TypeConstraint::new(
            node.clone(),
            visitor.block_type(self.output.clone()),
        ));
    }
}

impl Codegen for BlockTypeNode {
    fn codegen(&self, ctx: &mut CodegenCtx<'_>) -> Result<(), CodegenError> {
        Err(ctx.error())
    }
}
